Functoriality of sequential colimits (#919)
UniMath/agda-unimath: src/synthetic-homotopy-theory/functoriality-sequential-colimits.lagda.md

open import foundation.commuting-prisms-of-maps
open import foundation.commuting-squares-of-homotopies
open import foundation.commuting-squares-of-maps
open import foundation.dependent-pair-types
open import foundation.function-extensionality
open import foundation.homotopies
open import foundation.universe-levels

open import foundation-core.equivalences
open import foundation-core.function-types
open import foundation-core.functoriality-dependent-pair-types
open import foundation-core.whiskering-homotopies

open import synthetic-homotopy-theory.cocones-under-sequential-diagrams
open import synthetic-homotopy-theory.equivalences-sequential-diagrams
open import synthetic-homotopy-theory.morphisms-sequential-diagrams
open import synthetic-homotopy-theory.sequential-colimits
open import synthetic-homotopy-theory.sequential-diagrams
open import synthetic-homotopy-theory.universal-property-sequential-colimits
```

</details>

## Idea

A
[morphism of sequential diagrams](synthetic-homotopy-theory.morphisms-sequential-diagrams.md)
`f : (A, a) → (B, b)` induces a map `f∞ : A∞ → B∞` between the
[standard sequential colimits](synthetic-homotopy-theory.sequential-colimits.md)

> of the diagrams.

Author: Let me know if the new wording is better.

```agda
module _
  { l1 l2 l3 : Level} {A : sequential-diagram l1} {B : sequential-diagram l2}
  ( f : hom-sequential-diagram A B) {X : UU l3}
  where

  map-cocone-hom-sequential-diagram :
    cocone-sequential-diagram B X → cocone-sequential-diagram A X
  map-cocone-hom-sequential-diagram c =
    comp-hom-sequential-diagram A B (constant-sequential-diagram X) c f
```

### A morphism of sequential diagrams induces a map of sequential colimits

The induced map

```text
        a₀      a₁      a₂
    A₀ ---> A₁ ---> A₂ ---> ⋯ ---> A∞
    |       |       |              |
 f₀ |       | f₁    | f₂           | f∞
    V       V       V              V
    B₀ ---> B₁ ---> B₂ ---> ⋯ ---> B∞
        b₀      b₁      b₂
```

then induces a cocone under `(A, a)` with codomain `B∞`, which is homotopic to
the standard cocone under `(B, b)` precomposed by `f`.

This homotopy provides
[vertical commuting prisms of maps](foundation.commuting-prisms-of-maps.md),

```text
          Aₙ₊₁
         ^| \
       /  |   \
     /    |fₙ₊₁ V
    Aₙ ---------> A∞
    |     |      |
    |     V      |
 fₙ |     Bₙ₊₁   | f∞
    |    ^  \    |
    |  /      \  |
    |/          V|

>     Bₙ --------> B∞ ,

Author: ![20231203 151206 screen](https://github.com/UniMath/agda-unimath/assets/15523887/320b2cdc-9d81-4f5b-8616-897116826cf2)

This one actually renders fine for me though?

          ( inv-htpy
            ( htpy-htpy-cocone-map-hom-standard-sequential-colimit C g
              ( succ-ℕ n)))
          ( naturality-map-hom-sequential-diagram C g n)
          ( coherence-triangle-cocone-standard-sequential-colimit n)
          ( prism-htpy-cocone-map-hom-standard-sequential-colimit B f n)
          ( prism-htpy-cocone-map-hom-standard-sequential-colimit C g
            ( n))))) ∙h
    ( inv-htpy-assoc-htpy
      ( htpy-htpy-cocone-map-hom-standard-sequential-colimit C
        ( comp-hom-sequential-diagram A B C g f)
        ( n))
      ( _)
      ( ( ( map-hom-standard-sequential-colimit C g) ∘
          ( map-hom-standard-sequential-colimit B f)) ·l
        ( coherence-triangle-cocone-standard-sequential-colimit n))))

  preserves-comp-map-hom-standard-sequential-colimit :
    ( map-hom-standard-sequential-colimit C
      ( comp-hom-sequential-diagram A B C g f)) ~
    ( ( map-hom-standard-sequential-colimit C g) ∘
      ( map-hom-standard-sequential-colimit B f))
  preserves-comp-map-hom-standard-sequential-colimit =
    htpy-htpy-out-of-standard-sequential-colimit A
      htpy-preserves-comp-map-hom-standard-sequential-colimit
```

> ### An equivalence of sequential diagrams induces an equivalence of cocones

Author: I'm not sure off the top of my head. I'll look into it, but I think that I don't need it at the moment, so if I can't get it done by tonight I'll open an issue for it and leave it at that.

> ### An equivalence of sequential diagrams induces an equivalence of cocones

Reviewer: That's fair :) If you don't need it then don't stress it. I'm asking because this lemma seems useful for #930 and #842.

Additionally, the inverse of the induced equivalence is the map induced by the

> inverse equivalence of sequential diagrams, i.e. `(e∞)⁻¹ ≐ (e⁻¹)∞`.

Author: Definitional equality - I'm pretty sure it's used in other places in the prose, but I'm not at my computer to check rn

```agda
module synthetic-homotopy-theory.functoriality-sequential-colimits where
```

<details><summary>Imports</summary>

```agda
open import elementary-number-theory.natural-numbers

open import foundation.action-on-identifications-functions
open import foundation.commuting-prisms-of-maps
open import foundation.commuting-squares-of-homotopies
open import foundation.commuting-squares-of-maps
open import foundation.dependent-pair-types
open import foundation.function-extensionality
open import foundation.homotopies
open import foundation.universe-levels

open import foundation-core.equivalences
open import foundation-core.function-types
open import foundation-core.functoriality-dependent-pair-types
open import foundation-core.whiskering-homotopies

open import synthetic-homotopy-theory.cocones-under-sequential-diagrams
open import synthetic-homotopy-theory.equivalences-sequential-diagrams
open import synthetic-homotopy-theory.morphisms-sequential-diagrams
open import synthetic-homotopy-theory.sequential-colimits
open import synthetic-homotopy-theory.sequential-diagrams
open import synthetic-homotopy-theory.universal-property-sequential-colimits
```

</details>

## Idea

> Taking a [sequential colimit](synthetic-homotopy-theory.sequential-colimits.md)
> of a [sequential diagram](synthetic-homotopy-theory.sequential-diagrams.md) is a
> functorial action `(A, a) ↦ A∞`. In particular, a
> [morphism of sequential diagrams](synthetic-homotopy-theory.morphisms-sequential-diagrams.md)
> `f : (A, a) → (B, b)` induces a map `f∞ : A∞ → B∞` between the
> [standard sequential colimits](synthetic-homotopy-theory.sequential-colimits.md)
> of the diagrams, and this action preserves the identity morphism and morphism
> composition.

Reviewer: Since you're explaining functoriality in its entirety I would drop "in particular":
```suggestion
Taking the [sequential colimit](synthetic-homotopy-theory.sequential-colimits.md)
of a [sequential diagram](synthetic-homotopy-theory.sequential-diagrams.md) is a
functorial action `(A, a) ↦ A∞`. So a
[morphism of sequential diagrams](synthetic-homotopy-theory.morphisms-sequential-diagrams.md)
`f : (A, a) → (B, b)` induces a map `f∞ : A∞ → B∞` between the
[standard sequential colimits](synthetic-homotopy-theory.sequential-colimits.md)
of the diagrams, and this action preserves the identity morphism and morphism
composition.
```

              ( succ-ℕ n)))
          ( naturality-map-hom-sequential-diagram C g n)
          ( coherence-triangle-cocone-standard-sequential-colimit n)
          ( prism-htpy-cocone-map-hom-standard-sequential-colimit B f n)
          ( prism-htpy-cocone-map-hom-standard-sequential-colimit C g
            ( n))))) ∙h
    ( inv-htpy-assoc-htpy
      ( htpy-htpy-cocone-map-hom-standard-sequential-colimit C
        ( comp-hom-sequential-diagram A B C g f)
        ( n))
      ( _)
      ( ( ( map-hom-standard-sequential-colimit C g) ∘
          ( map-hom-standard-sequential-colimit B f)) ·l
        ( coherence-triangle-cocone-standard-sequential-colimit n))))

  preserves-comp-map-hom-standard-sequential-colimit :
    ( map-hom-standard-sequential-colimit C
      ( comp-hom-sequential-diagram A B C g f)) ~
    ( ( map-hom-standard-sequential-colimit C g) ∘
      ( map-hom-standard-sequential-colimit B f))
  preserves-comp-map-hom-standard-sequential-colimit =
    htpy-htpy-out-of-standard-sequential-colimit A
      htpy-preserves-comp-map-hom-standard-sequential-colimit
```

### An equivalence of sequential diagrams induces an equivalence of cocones

> Additionally, the underlying map of the inverse equivalence if definitionally

Reviewer: ```suggestion
Additionally, the underlying map of the inverse equivalence is definitionally
```
